### PR TITLE
Handle IForToLoopOperation over user-defined types in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >Clone [](start = 86, length = 5)

Are these nodes nested inside the ConditionOpt? ie it's not the condition itself, but the children of the condition?

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > Are these nodes nested inside the ConditionOpt? ie it's not the condition itself, but the children of the condition?

Yes, they are the Values of the case clauses and are also used as operands of binary operators inside the condition.

---
In reply to: [187495725](https://github.com/dotnet/roslyn/pull/26771#discussion_r187495725) [](ancestors = 187495725)

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > This sub-tree or nodes from it are not available to regular consumers, it is internal and used only by CFG builder.

Thanks for the clarification. Can you please add this to the comment, or comment on Clone about this?

**Review by 333fred — APPROVED** [review]
LGTM modulo updating documentation to state what the expectations on calling Clone are.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: > Can you please add this to the comment, or comment on Clone about this?

Sure, I'll add this information to the comment above in the next PR. 

---
In reply to: [187683676](https://github.com/dotnet/roslyn/pull/26771#discussion_r187683676) [](ancestors = 187683676)

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need a second review

File: src/Compilers/Core/Portable/SpecialTypeExtensions.cs
Comment (cston) [reply src/Compilers/Core/Portable/SpecialTypeExtensions.cs]: It looks the method was added in the last check in.

---
In reply to: [187695534](https://github.com/dotnet/roslyn/pull/26771#discussion_r187695534) [](ancestors = 187695534,187694506)

**Review by mavasani — APPROVED** [review]
:shipit:

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >Referencefor [](start = 46, length = 12)

`ReferenceFor`

Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >reevaluate it again [](start = 158, length = 19)

Perhaps `evaluate it again` or simply `reevaluate it`, here and other instances below.

File: src/Compilers/Core/Portable/SpecialTypeExtensions.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/SpecialTypeExtensions.cs]: > It looks the method was added in the last check in.

It was moved from VB to the shared layer. The intent was to not change implementation to much in the process and the original code was in terms of PrimitiveTypeCode

---
In reply to: [187697482](https://github.com/dotnet/roslyn/pull/26771#discussion_r187697482) [](ancestors = 187697482,187695534,187694506)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > ReferenceFor 

Will adjust the name in the next PR

---
In reply to: [187707603](https://github.com/dotnet/roslyn/pull/26771#discussion_r187707603) [](ancestors = 187707603)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Perhaps evaluate it again or simply reevaluate it, here and other instances below.

Will make this adjustment in the next PR

---
In reply to: [187709105](https://github.com/dotnet/roslyn/pull/26771#discussion_r187709105) [](ancestors = 187709105)

**Review by cston — APPROVED** [review]

**Pull request merged**
